instrumentation/grpc/src/main/java/brave/grpc/TraceContextBinaryMarshaller.java

```diff
@@ -0,0 +1,111 @@
+package brave.grpc;
+
+import brave.propagation.TraceContext;
+import io.grpc.Metadata.BinaryMarshaller;
+import java.util.logging.Logger;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * This logs instead of throwing exceptions.
+ *
+ * <p>See https://github.com/census-instrumentation/opencensus-specs/blob/master/encodings/BinaryEncoding.md
+ */
+final class TraceContextBinaryMarshaller implements BinaryMarshaller<TraceContext> {
+  static final Logger logger = Logger.getLogger(TraceContextBinaryMarshaller.class.getName());
+  static final byte
+      VERSION = 0,
+      TRACE_ID_FIELD_ID = 0,
+      SPAN_ID_FIELD_ID = 1,
+      TRACE_OPTION_FIELD_ID = 2;
+
+  private static final int FORMAT_LENGTH =
+      4 /* version + 3 fields */ + 16 /* trace ID */ + 8 /* span ID */ + 1 /* sampled bit */;
+
+  @Override public byte[] toBytes(TraceContext traceContext) {
+    checkNotNull(traceContext, "traceContext");
+    byte[] bytes = new byte[FORMAT_LENGTH];
+    bytes[0] = VERSION;
+    bytes[1] = TRACE_ID_FIELD_ID;
+    writeLong(bytes, 2, traceContext.traceIdHigh());
+    writeLong(bytes, 10, traceContext.traceId());
+    bytes[18] = SPAN_ID_FIELD_ID;
+    writeLong(bytes, 19, traceContext.spanId());
+    bytes[27] = TRACE_OPTION_FIELD_ID;
+    if (traceContext.sampled() != null && traceContext.sampled()) {
+      bytes[28] = 1;
+    }
+    return bytes;
+  }
+
+  @Override public TraceContext parseBytes(byte[] bytes) {
+    if (bytes == null) throw new NullPointerException("bytes == null"); // programming error
+    if (bytes.length == 0 || bytes[0] != VERSION) {
+      logger.fine("Unsupported version.");
+      return null;
+    }
+    if (bytes.length < FORMAT_LENGTH - 2 /* sampled field + bit is optional */) {
+      logger.fine("Invalid input: truncated");
+      return null;
+    }
+    long traceIdHigh, traceId, spanId;
+    Boolean sampled = null;
+    int pos = 1;
+    if (bytes[pos] == TRACE_ID_FIELD_ID) {
+      pos++;
+      traceIdHigh = readLong(bytes, pos);
+      traceId = readLong(bytes, pos + 8);
+      pos += 16;
+    } else {
+      logger.fine("Invalid input: expected trace ID at offset " + pos);
+      return null;
+    }
+    if (bytes[pos] == SPAN_ID_FIELD_ID) {
+      pos++;
+      spanId = readLong(bytes, pos);
+      pos += 8;
+    } else {
+      logger.fine("Invalid input: expected span ID at offset " + pos);
```

Comment:
self-note: kill all these string cats as they shouldn't happen unless fine is enabled